### PR TITLE
Child chain

### PR DESCRIPTION
**Pull request by bun919tw**
issue #34 

Use web3 in child chain to communicate with ganache-cli for deposit and block submitting.

### CODE REVIEW

**Comment by kfichter** [comment]
Changes are good :+1:, but requesting changes -

Tests don't pass because we're currently initializing child chain in tests as:

```
child_chain = ChildChain(None, None, None)
```

However, now that we're using web3 in `child_chain.py`, we'll need to add a provider. The simplest fix is to initialize `child_chain` as:

```
child_chain = ChildChain(None)
```

assuming we have `ganache-cli` running on `localhost:8545`.

However, it's likely that users will want to test this project on their local machines without shutting down their testnet/mainnet local nodes. A few potential solutions:
- Allow users to specify a port when testing
- Add `ganache-cli` as a dev dependency and host on an open port for the duration of testing

I'd also request that tests be added to cover `apply_deposit` and `submit_block`.

**Comment by bun919tw** [comment]
@kfichter Thanks for the feedback. I've made some modifications.

Sorry, I'm not skilled in unittest. I ended up launching a ganache-cli in a given port when testing. All the tests pass now locally but travis is not happy because it doesn't recognize ganache-cli. Please give me some feedback so I can approve it! Thanks.

**Comment by kfichter** [comment]
I think having users launch their own Ethereum node and then adding some sort of optional flag to specify the port is the best solution here. This allows users to test using whatever Ethereum implementation they desire. In that respect, I think we shouldn't run `ganache-cli` while testing (going against my initial recommendation). 

I'm going to add an issue to change the travis build so that we're running an instance of `ganache-cli` and exposing the RPC API on the default port. This should solve the issue temporarily, but I request that 

```
@pytest.fixture
  def child_chain(root_chain):
      child_chain = ChildChain(AUTHORITY, HTTPProvider('http://localhost:{}'.format(GANACHE_TEST_PORT)), None)
```

be modified so that `GANACHE_TEST_PORT` is a CLI flag that defaults to `8545` and that

```
 @pytest.fixture
 def root_chain():
     subprocess.Popen(['ganache-cli', '-m=plasma_mvp', '-p={}'.format(GANACHE_TEST_PORT)])
```

be reverted entirely.

**Comment by bun919tw** [comment]
@kfichter 
I let the ganache-cli port to be a CLI argument when testing. Is that something you said? Thanks!

**Comment by bun919tw** [comment]
Hey, @kfichter 
Finally~ all checks have passed.
Here are my thoughts. I mock all the root chain reference in `test_child_chain.py` because I only want to test the function correctness in child chain. Compare to provide a real testrpc, I think this is more in line with the unittest logic. However, I do think launching a testrpc when testing is necessary, because we might write integration tests in the future.
Does this make sense to you?

File: tests/child_chain/test_child_chain.py [hunk 1]
```diff
@@ -1,26 +1,31 @@
...
 @pytest.fixture
-def root_chain(get_contract):
-    return get_contract('RootChain/RootChain.sol')
-
-
-@pytest.fixture
-def child_chain(root_chain):
-    child_chain = ChildChain(None, None, None)
+def child_chain():
+    with patch('plasma.child_chain.child_chain.ChildChain.__init__', return_value=None) as child_chain:
```
Comment (boolafish) [inline tests/child_chain/test_child_chain.py]: I guess we don't need the patch and 
```python
child_chain = ChildChain(None, None, None)
child_chain.root_chain = Mock()
```
should be enough?

From what I read is that you patch the `__init__` function as `child_chain`, but soon make `child_chain` become `ChildChain(None, None, None)`, so the patch is probably not used.
And since probably we don't need any magic value from the `root_chain`, `Mock` should probably be enough.

File: tests/child_chain/test_child_chain.py [hunk 2]
```diff
@@ -53,22 +58,72 @@ def test_send_tx_invalid_sig(child_chain):
...
+        child_chain.submit_block(block)
+
+
+def test_submit_block_invalid_sig(child_chain):
+    block = child_chain.current_block
+    block.make_mutable()
+    block.sign(invalid_block_key)
+    block = rlp.encode(block).hex()
+
+    with pytest.raises(AssertionError):
```
Comment (boolafish) [inline tests/child_chain/test_child_chain.py]: Minor: I would recommend to assert some message as well in submit_block
eg.
```python
assert valid_signature, 'invalid signature'
```

then here can be
```python
with pytest.raises(AssertionError) as e:
    child_chain.submit_block(block)
assert str(e.value) == 'invalid signature'
```
So generally we can make sure we are testing on the assert that we really want, and not from other places. (Though probably not the case here)

Or, personally I prefer to have custom Exceptions rather than assert. But it seems like the package have the convention of using assert, so I am not sure moving to custom Exception would be better or not.

Comment (kfichter) [reply tests/child_chain/test_child_chain.py]: I don't think we're set on the usage of `AssertionError`, I agree that it's probably better to throw custom errors. Maybe something like `InvalidBlockSignatureError`? Is that too wordy? It's relatively easy to implement these as:
```
class InvalidBlockSignatureError(Exception):
    """Raise when a block is signed by anyone but the operator"""
```

**Comment by kfichter** [comment]
@bun919tw Yes, mocking the root chain in `test_child_chain.py` makes 100% sense to me. I think it's fine to keep `ganache-cli` in the travis build (for now) because we'll need it for `test_root_chain.py` eventually.

**Comment by kfichter** [comment]
This is an artifact of the original `submit_block`, but we should probably consider validating that the block being submitted here (`block`):

https://github.com/bun919tw/plasma-mvp/blob/a49651a0e5dc513a01da387c11cbda5e6c894241/plasma/child_chain/child_chain.py#L83-L92

is equivalent to `current_block`. Currently, we aren't checking that this is the case. I don't think we want an operator to be able to submit a `block` that isn't the `current_block`. Adding this check should also remove the necessity of checking that all of the transactions in the block are valid, because we'll already have checked that in `apply_transaction` (which should be the only way to append to the current block's transaction set).

**Comment by bun919tw** [comment]
Add two more commits.

- Discussed with @boolafish, I modified the ChildChain's constructor. Pass in a root chain instance as a parameter instead of a root chain provider.
- Advised from @kfichter, I added the validation of the submitting block content and its corresponding test.

**Comment by kfichter** [comment]
I agree with customizing exceptions in another PR, I'll go ahead and open an issue for that right now. 

Something I didn't think of earlier:

I think `ChildChain` needs to be modified in case it receives an event late. To illustrate, let's assume `deposit_filter.watch(self.apply_deposit)` fails and `self.apply_deposit` is being called late or not at all. 

This has some potential downsides:
1. The operator never sees the deposit, so the user can't spend their deposit and must exit. Or the operator sees the deposit late and the user can't spend until the operator has seen the deposit. I'm not really worried about this, it probably won't be a huge issue.
2. The operator sees the deposit late and applies the deposit via:
```
...
        deposit_tx = Transaction(0, 0, 0, 0, 0, 0,
                                 newowner1, amount1, b'\x00' * 20, 0, 0)
        deposit_block = Block([deposit_tx])
        # Add block validation
        self.blocks[self.current_block_number] = deposit_block
        self.current_block_number += 1
```

However, `current_block_number` might not be `currentChildBlock` as defined in `RootChain.deposit()`:

```
        childChain[currentChildBlock] = childBlock({
            root: root,
            created_at: block.timestamp
        });
        currentChildBlock = currentChildBlock.add(1);
        Deposit(txList[6].toAddress(), txList[7].toUint());
```

This means that block numbers between `ChildChain` and the root chain contract will differ. This is a serious problem because `ChildChain.get_block(blknum)` will return the wrong block. 

I need to think about potential mitigations for this, just a second. The issue gets considerably more complex in the case that the operator is on a chain that experiences a reorg.

**Comment by DavidKnott** [comment]
LGTM, @bun919tw Thanks a lot for the contribution!
